### PR TITLE
Feat: AddDataValidation() and DeleteDataValidation() are concurrency safe, and add some unit test and code optimization.

### PR DESCRIPTION
**Pull request by JackMin1314**
# PR Details

This pr contains two commits, non-breaking change which fixes an issue and adds a function.

1. **fix**: data validation is unsafe when used in concurrent condition[#1825](https://github.com/qax-os/excelize/issues/1825)
2. **feat**: supports batch append data validation to the worksheet and optimse code[#1826](https://github.com/qax-os/excelize/issues/1826)

## Description

1. For the first commit which open related issue #1825 and has detail description about this. Use the mutex's method Lock() and Unlock() which is in the **\*File** object.

2. Optimse code: Extract some code about xlsxDataValidation creation and keep the contents of the AddDataValidation function clean
3. And  a  new function named **BatchAddDataValidation()**  to batch append data validation to the worksheet for some typical excel data validation template scenario.

## Related Issue

[#1825](https://github.com/qax-os/excelize/issues/1825)

[#1826](https://github.com/qax-os/excelize/issues/1826)

## How Has This Been Tested

There are some test cases already written in test file *datavalidation_test.go* and two testing functions: **TestConcurrentAddDataValidation()** for high concurrency and  **TestBatchAddDataValidation()** to batch append data validation.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: datavalidation.go [hunk 1]
```diff
@@ -256,30 +281,33 @@ func (f *File) AddDataValidation(sheet string, dv *DataValidation) error {
 	if err != nil {
 		return err
 	}
+	f.mu.Lock()
```
Comment (xuri) [inline datavalidation.go:284]: I suggest using the embed mutex in worksheet struct:

```go
ws.mu.Lock()
defer ws.mu.Unlock()
```

File: datavalidation.go [hunk 2]
```diff
@@ -256,30 +281,33 @@ func (f *File) AddDataValidation(sheet string, dv *DataValidation) error {
...
+	defer f.mu.Unlock()
```
Comment (xuri) [inline datavalidation.go:285]: All concurrency safety functions have a note in the documentation. The AddDataValidation function is not concurrency safety currently, so these changes should be a feature but not a bug, and I certainly accept this feature proposal. I suggest adding a note in the documentation for this function:

```go
// AddDataValidation provides set data validation on a range of the worksheet
// by given data validation object and worksheet name. This function is 
// concurrency safe. The data validation object can be created by
// NewDataValidation function.
```

File: datavalidation.go (hunk 1 above)
Comment (xuri) [inline datavalidation.go:284]: I suggest using the embed mutex in worksheet struct:

```go
ws.mu.Lock()
defer ws.mu.Unlock()
```

File: datavalidation_test.go [hunk 3]
```diff
@@ -106,6 +109,124 @@ func TestDataValidation(t *testing.T) {
 	assert.Equal(t, []*DataValidation(nil), dataValidations)
 }
 
+func TestConcurrentAddDataValidation(t *testing.T) {
```
Comment (xuri) [inline datavalidation_test.go:112]: All concurrency test should be add to `TestConcurrency` function:

```diff
diff --git a/cell_test.go b/cell_test.go
index 0ed6e87..78c1ee7 100644
--- a/cell_test.go
+++ b/cell_test.go
@@ -88,6 +88,12 @@ func TestConcurrency(t *testing.T) {
 			visible, err := f.GetColVisible("Sheet1", "A")
 			assert.NoError(t, err)
 			assert.Equal(t, true, visible)
+			// Concurrency add data validation
+			dv := NewDataValidation(true)
+			dv.Sqref = fmt.Sprintf("A%d:B%d", val, val)
+			dv.SetRange(10, 20, DataValidationTypeWhole, DataValidationOperatorGreaterThan)
+			dv.SetInput(fmt.Sprintf("title:%d", val), strconv.Itoa(val))
+			f.AddDataValidation("Sheet1", dv)
 			wg.Done()
 		}(i, t)
 	}
@@ -97,6 +103,10 @@ func TestConcurrency(t *testing.T) {
 		t.Error(err)
 	}
 	assert.Equal(t, "1", val)
+	// Test the length of data validation
+	dataValidations, err := f.GetDataValidations("Sheet1")
+	assert.NoError(t, err)
+	assert.Len(t, dataValidations, 5)
 	assert.NoError(t, f.SaveAs(filepath.Join("test", "TestConcurrency.xlsx")))
 	assert.NoError(t, f.Close())
 }
```

so this test function can be removed, and we can test it with command `go test -v -race --run TestConcurrency`.

File: datavalidation.go [hunk 4]
```diff
@@ -256,30 +281,33 @@ func (f *File) AddDataValidation(sheet string, dv *DataValidation) error {
...
-		ShowInputMessage: dv.ShowInputMessage,
-		Sqref:            dv.Sqref,
-		Type:             dv.Type,
+	dataValidation := newXlsxDataValidation(dv)
+	ws.DataValidations.DataValidation = append(ws.DataValidations.DataValidation, dataValidation)
+	ws.DataValidations.Count = len(ws.DataValidations.DataValidation)
+	return err
+}
+
+// BatchAddDataValidation is a function supports batch append data validation to the worksheet, reference to AddDataValidation().
```
Comment (xuri) [inline datavalidation.go:295]: Any performance benchmark for this? To keep the library core and easy, and follow the "typically one right way to do most things" principles, anyone can easily make a custom function to add multiple data validations by themselves: batch append data validations, batch append tables, etc., so I don't suggest adding these functions.

**Review by xuri — CHANGES_REQUESTED** [review]
Thanks for your pull request. I've left some comments.

File: datavalidation.go (hunk 2 above)
Comment (JackMin1314) [reply datavalidation.go]: Thanks for your suggestion. I have added in the new  commit.

File: datavalidation_test.go (hunk 3 above)
Comment (JackMin1314) [reply datavalidation_test.go]: Thanks for your suggestion. I have added in the `TestConcurrency` function and remove `TestConcurrentAddDataValidation()` which in datavalidation_test.go

File: datavalidation.go (hunk 4 above)
Comment (JackMin1314) [reply datavalidation.go]: **BatchAddDataValidation() is not added in the latest commit.**  Here are some results of the benchmark.
BatchAddDataValidation() reduces the average op time by 37.9% compared to AddDataValidation(), but has an additional memory footprint of 4.3% and 1 alloc. 

example:
go test -benchmem -run=^$ -bench DataValidation$  -benchtime=100000x -cpu=8
goos: linux
goarch: amd64
pkg: github.com/xuri/excelize/v2
cpu: 11th Gen Intel(R) Core(TM) i5-11320H @ 3.20GHz
BenchmarkAddDataValidation-8              100000             33783 ns/op           20920 B/op        300 allocs/op
BenchmarkBatchAddDataValidation-8         100000             19289 ns/op           21815 B/op        301 allocs/op
PASS
ok      github.com/xuri/excelize/v2     5.465s

**Review by xuri — APPROVED** [review]
LGTM, thanks for your contribution.

**Pull request merged**